EDK2-BUG4244 V2 - Print out the PCIe device path for SCT failures.

PrachotanReddy opened the pull request:
Modified DevicePath.c to include a wrapper function for printing device path. All you need is SctPrintDevicePathFromProtocol function. Tests Modified:
AdapterInfoBBTest
HIIConfigAccessBBTest
PxeBaseCodeBBTest
SimpleNetworkBBTest

UEFI-SCT
Cc: G Edhaya Chandran <Edhaya.Chandran@arm.com>
Cc: Barton Gao <gaojie@byosoft.com.cn>
Cc: Carolyn Gjertsen <Carolyn.Gjertsen@amd.com>
Cc: Samer El-Haj-Mahmoud <Samer.El-Haj-Mahmoud@arm.com>
Cc: Eric Jin <eric.jin@intel.com>
Cc: Arvin Chen <arvinx.chen@intel.com>
Cc: Supreeth Venkatesh <Supreeth.Venkatesh@amd.com>

sunnywang-arm commented on uefi-sct/SctPkg/Include/SctLib.h:
```diff
@@ -132,6 +132,16 @@ SctDevicePathStrFromProtocol (
   IN EFI_GUID    *Guid
   );
 
+struct _EFI_STANDARD_TEST_LIBRARY_PROTOCOL;
```
There should be no need to duplicate these two lines for the EFI_STANDARD_TEST_LIBRARY_PROTOCOL. 
Could you add gEfiStandardTestLibraryGuid to SctLib.inf instead and see if it works? If not, can we directly include "Protocol/StandardTestLibrary.h"?

PrachotanReddy replied:
Hi Sunny
Adding the StandardTestLibrary header file is causing redeclaration issues, hence I included it in devicepath.c and made a forward declaration in sctlib.h

sunnywang-arm replied:
In this case, if we want to use the wrapper function, we need to figure out what is wrong with the StandardTestLibrary header file. 
Again, I'm fine with NOT using the wrapper function now as it is more complicated than I thought.

PrachotanReddy replied:
The issue, this was declared in multiple header files, tracing dependency was a bit challenging hence had to go with this approach:
typedef enum {
  EFI_TEST_ASSERTION_PASSED,
  EFI_TEST_ASSERTION_WARNING,
  EFI_TEST_ASSERTION_FAILED
} EFI_TEST_ASSERTION;
If we do not want to use a wrapper function we can roll back to the previous approach where we were only calling SctDevicePathStrFromProtocol and use recordmessage within the test function.

sunnywang-arm replied:
Agree, so let's roll back to the previous approach.

sunnywang-arm commented on uefi-sct/SctPkg/Library/SctLib/DevicePath.c:
```diff
@@ -16,7 +16,8 @@
 #include "SctLib.h"
-
+#include "Protocol/StandardTestLibrary.h"
+#include "Library/EfiTestLib.h"
```
Just curious why we need to include EfiTestLib.h?  
If we need to include EfiTestLib, please add it to SctLib.inf file instead.

PrachotanReddy replied:
EfiTestLib.h is needed for gtbs to use at `gtBS->FreePool (DevicePathStr)`
Similarly including it directly with sctlib is causing redeclaration issues, so I limited the scope of the inclusion by just including it in devicepath.c

sunnywang-arm replied:
In this case, you should add UefiBootServicesTableLib to SctLib.inf instead.

sunnywang-arm reviewed (changes requested):
Thanks for addressing my comments, @PrachotanReddy 
Few more comments

PrachotanReddy committed:
EDK2-Test BUG4244 V2 - Print PCIe device path for SCT failures.

Tests Modified:
AdapterInfoBBTest
HIIConfigAccessBBTest
PxeBaseCodeBBTest
SimpleNetworkBBTest
Uses SctDevicePathStrFromProtocol function from DevicePath.c

UEFI-SCT
Cc: G Edhaya Chandran <Edhaya.Chandran@arm.com>
Cc: Barton Gao <gaojie@byosoft.com.cn>
Cc: Carolyn Gjertsen <Carolyn.Gjertsen@amd.com>
Cc: Samer El-Haj-Mahmoud <Samer.El-Haj-Mahmoud@arm.com>
Cc: Eric Jin <eric.jin@intel.com>
Cc: Arvin Chen <arvinx.chen@intel.com>
Cc: Supreeth Venkatesh <Supreeth.Venkatesh@amd.com>
Signed-off-by: Prachotan Bathi <prachotan.bathi@arm.com>
